mne_bids/utils.py

```diff
@@ -29,6 +30,142 @@
+    prefix : str | None
+        The prefix for the filename to be created. E.g., a path to the folder
+        in which you wish to create a file with this name.
+    suffix : str | None
+        The suffix for the filename to be created. E.g., 'audio.wav'.
+
+    Returns
+    -------
+    filename : str
+        The BIDS filename you wish to create.
+
+    Examples
+    --------
+    >>> bids_basename = make_bids_basename(subject='test', session='two', task='mytask', suffix='data.csv')  # noqa: E501
+    >>> print(bids_basename)
+    sub-test_ses-two_task-mytask_data.csv
+    >>> print(bids_basename.__repr__())
+    BIDSPath (sub-test_ses-two_task-mytask_data.csv)
+    """
+
+    def __init__(self, sub=None, ses=None, task=None,
+                 acq=None, run=None, proc=None,
+                 rec=None, space=None,
+                 prefix=None, suffix=None):
+        self.sub = sub
+        self.ses = ses
+        self.task = task
+        self.acq = acq
+        self.run = run
+        self.proc = proc
+        self.rec = rec
+        self.space = space
+        self.prefix = prefix
+        self.suffix = suffix
+
+        # run string representation to check validity of arguments
+        self._get_name()
+
+    @property
+    def entities(self):
+        """Return dictionary of the BIDS entities."""
+        keys = ('sub', 'ses', 'task', 'acq',
+                'proc', 'acq', 'run', 'rec',
+                'space', 'suffix', 'prefix')
+        entities = OrderedDict()
+
+        for key in keys:
+            value = getattr(self, key, None)
+            if value is not None:
+                entities[key] = value
+        return entities
+
+    def __setitem__(self, key, value):
+        """Set item as a dictionary, and perform validation checks."""
+        if key not in ('sub', 'ses', 'task', 'acq',
+                       'proc', 'acq', 'run', 'rec',
+                       'space', 'suffix', 'prefix'):
+            raise ValueError('Key must be one of blah, got %s' % key)
+        setattr(self, key, value)
+
+    def __str__(self):
+        """Return the string representation of the path."""
+        return self._get_name()
+
+    def __repr__(self):
+        """Representation in the style of `pathlib.Path`."""
+        return "{}({!r})".format(self.__class__.__name__, self._get_name())
+
+    def __fspath__(self):
+        """Return the string representation for any fs functions."""
+        return str(self)
+
+    def __bytes__(self):
+        """Return the bytes representation of the path.
+
+        This is only recommended to use under Unix.
+        """
+        return os.fsencode(str(self))
+
+    def __eq__(self, other):
+        """Compare str representations."""
+        return str(self) == str(other)
+
+    def startswith(self, prefix, start=None, end=None):
+        """Return True/False for the str startswith."""
+        return str(self).startswith(prefix, start=start, end=end)
```

Comment:
```suggestion
```

I would say don't mix `BIDSPath` with `str` or other kinds of objects. The class should be as thin as possible